### PR TITLE
WINC-498: [wmco] vSphere README updates

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Manual steps for vSphere for configuring WMCO.

cc @aravindhp @gmarkley-VI @jcpowermac @VaishnaviHire

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -213,6 +213,93 @@ oc apply -f deploy/olm-catalog/subscription.yaml
...
 ### Creating a Windows MachineSet
+
+#### Pre-configuration steps in case of vSphere cluster:
+##### Windows VM assumptions:
+- We tested using updated version of Windows Server 1909 VM image and corresponding container images
+- We assume that the Windows image is configured with SSH so that WMCO can access it. Following are steps to comfigure SSH
+  - Configure SSH on the Windows VM using the [powershell script](https://gist.githubusercontent.com/ravisantoshgudimetla/087bf49088f70edf7f97777386d12f18/raw/0f13048697130a0265ed1050591361fdebcee3ae/powershell.yaml)
+  - Create a new file called C:\Users\Administrator\.ssh\authorized_keys and copy your public key to this file
+- We assume that the VMWare tools has been installed on the Windows VM
+- We assume that the tools.conf at C:\ProgramData\VMware\VMware Tools\tools.conf is having an entry for
```
Comment (jcpowermac) [inline README.md:224]: Documentation about configuration file location:
https://docs.vmware.com/en/VMware-Tools/11.1.0/com.vmware.vsphere.vmwaretools.doc/GUID-EA16729B-43C9-4DF9-B780-9B358E71B4AB.html

Maybe some text about if ip addresses are not available in vCenter for the guest check to make sure the service is running correct?

File: README.md [hunk 2]
```diff
@@ -213,6 +213,93 @@ oc apply -f deploy/olm-catalog/subscription.yaml
...
+  -`exclude-nics=`
+  
+  This is to ensure that tools won't ignore cloned the vNIC generated on the Windows VM by hybrid-overlay and the vNIC
+  shows in the vCenter console
+- We assume that the Windows image has been sysprep'ed and it is available for cloning:
+`C:\Windows\System32\Sysprep\sysprep.exe /generalize /oobe /shutdown`
```
Comment (jcpowermac) [inline README.md:230]: There has to be a `unattend.xml`, this will not work correctly as stated.

Comment (ravisantoshgudimetla) [reply README.md]: @jcpowermac  - Updated, ptal

File: README.md [hunk 3]
```diff
@@ -214,6 +214,94 @@ oc apply -f deploy/olm-catalog/subscription.yaml
...
+  - Create a new file called C:\Users\Administrator\.ssh\authorized_keys and copy your public key to this file
+- We assume that the VMWare tools has been installed on the Windows VM
+- We assume that the [tools.conf](https://docs.vmware.com/en/VMware-Tools/11.1.0/com.vmware.vsphere.vmwaretools.doc/GUID-EA16729B-43C9-4DF9-B780-9B358E71B4AB.html) at C:\ProgramData\VMware\VMware Tools\tools.conf is having an entry for
+  -`exclude-nics=`
+  
+ - This is to ensure that tools won't ignore the cloned vNIC generated on the Windows VM by hybrid-overlay and the VM has an IP Address in the vCenter.
+ Also, please ensure that the tools Windows service is running in the Windows VM.
+
+- We assume that the Windows image has been sysprep'ed and it is available for cloning using the following command:
+`C:\Windows\System32\Sysprep\sysprep.exe /generalize /oobe /shutdown /unattend:<path_to_unattend.xml>`
```
Comment (gmarkley-VI) [inline README.md:232]: Do you have an example of unattend.xml?

File: README.md [hunk 4]
```diff
@@ -214,6 +214,94 @@ oc apply -f deploy/olm-catalog/subscription.yaml
...
+
+- Generally, ignition config files are available on the internal API Server URL in case of IPI however in case of 
+vSphere, the ignition files are available on both internal and external API Server URLs. WMCO as of now, downloads the
+ignition files from internal API server. So, add a new DNS entry for `api-int.<cluster-name>.domain.com` which points
+to external API server URL `api.<cluster-name>.domain.com`. The external api should have been created as part of 
+[cluster install](https://docs.openshift.com/container-platform/4.5/installing/installing_vsphere/installing-vsphere-installer-provisioned.html)
+We will remove this requirement in future
+
+vSphere Windows VMs are cloned off an existing template or VM, for every VM following steps need to be done. We assume sysprep has been run on the VM:
+- Change the hostname of the Windows VM to match Windows Machine you created, following powershell command can be used:
```
Comment (LorbusChris) [inline README.md:243]:
```suggestion
- Change the hostname of the Windows VM to match Windows Machine you created. The following powershell command can be used to do so:
```

**Comment by jcpowermac** [comment]
@ravisantoshgudimetla I think the only thing left from my perspective is the `unattend.xml` example.

File: README.md [hunk 5]
```diff
@@ -214,6 +214,94 @@ oc apply -f deploy/olm-catalog/subscription.yaml
 ```
 
 ### Creating a Windows MachineSet
+
+#### Pre-configuration steps in case of vSphere cluster:
```
Comment (sebsoto) [inline README.md:218]: I dont think this should be written out here. I think it should be broken out into a seperate doc linked to under the https://github.com/openshift/windows-machine-config-operator/blob/master/README.md#pre-requisites section

File: README.md [hunk 6]
```diff
@@ -214,6 +214,94 @@ oc apply -f deploy/olm-catalog/subscription.yaml
...
+We will remove this requirement in future
+
+vSphere Windows VMs are cloned off an existing template or VM, for every VM following steps need to be done. We assume sysprep has been run on the VM:
+- Change the hostname of the Windows VM to match Windows Machine you created, following powershell command can be used:
+```
+Rename-Computer -NewName <machine_name> -Force -Restart
+```
+We are trying to automate the above steps and check this page to get latest updates.
+
+Sample vSphere MachineSet:
```
Comment (sebsoto) [inline README.md:249]: In a previous design meeting we decided that we should have a single example MachineSet within the README and link out to different docs for other providers. I think we decided to have vSphere be that single example. In that case we should move the azure example out of the doc and into a seperate doc file, like we did the AWS example

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
 ## Pre-requisites
 * OKD/OCP 4.6 cluster running on Azure or AWS, configured with [hybrid OVN Kubernetes networking](docs/setup-hybrid-OVNKubernetes-cluster.md)
```
Comment (aravindhp) [inline README.md:13]: `vSphere` needs to be added here.

File: README.md [hunk 8]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
 
+### Pre-configuration steps in case of vSphere cluster:
+#### Windows VM assumptions:
+- An updated version of Windows Server 1909 VM image and corresponding container images are used
+- The Windows image is configured with SSH so that WMCO can access it:
+  - SSH on the Windows VM is configured using the [powershell script](https://gist.githubusercontent.com/ravisantoshgudimetla/087bf49088f70edf7f97777386d12f18/raw/0f13048697130a0265ed1050591361fdebcee3ae/powershell.yaml)
```
Comment (aravindhp) [inline README.md:19]: Please add this to file to the repo.

File: README.md [hunk 9]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+to external API server URL `api.<cluster-name>.domain.com`. The external api should have been created as part of 
+[cluster install](https://docs.openshift.com/container-platform/4.5/installing/installing_vsphere/installing-vsphere-installer-provisioned.html)
+This requirement will be removed in the future.
+
+vSphere Windows VMs are cloned off an existing template or VM, for every VM the following steps need to be done. It is assumed that sysprep has been run on the VM:
+- Change the hostname of the Windows VM to match Windows Machine you created. The following powershell command can be used to do so:
+```
+Rename-Computer -NewName <machine_name> -Force -Restart
+```
+This process is going to be automated in the future. This document will be updated accordingly.
```
Comment (mansikulkarni96) [inline README.md:43]: This comment can be removed if docs are to be moved to the vsphere changes PR.

File: README.md [hunk 10]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
 ## Pre-requisites
 * OKD/OCP 4.6 cluster running on Azure or AWS, configured with [hybrid OVN Kubernetes networking](docs/setup-hybrid-OVNKubernetes-cluster.md)
 
+### Pre-configuration steps in case of vSphere cluster:
+#### Windows VM assumptions:
+- An updated version of Windows Server 1909 VM image and corresponding container images are used
+- The Windows image is configured with SSH so that WMCO can access it:
+  - SSH on the Windows VM is configured using the [powershell script](https://gist.githubusercontent.com/ravisantoshgudimetla/087bf49088f70edf7f97777386d12f18/raw/0f13048697130a0265ed1050591361fdebcee3ae/powershell.yaml)
+  - A new file called C:\Users\Administrator\.ssh\authorized_keys containing the public key to be used is created
```
Comment (aravindhp) [inline README.md:20]: Please apply italics to C:\Users\Administrator\.ssh\authorized_keys

File: README.md [hunk 11]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+- VMWare tools is installed, configured and running on the Windows VM:
+  - The [tools.conf](https://docs.vmware.com/en/VMware-Tools/11.1.0/com.vmware.vsphere.vmwaretools.doc/GUID-EA16729B-43C9-4DF9-B780-9B358E71B4AB.html) file at C:\ProgramData\VMware\VMware Tools\tools.conf has the following entry to ensure that the cloned vNIC generated on the Windows VM by hybrid-overlay won't be ignored and the VM has an IP Address in the vCenter
```
Comment (aravindhp) [inline README.md:22]: 120 col. Applies to all places where you are introducing long lines.

File: README.md [hunk 12]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+    - `exclude-nics=`
+  - The tools Windows service is running in the Windows VM.
```
Comment (aravindhp) [inline README.md:24]: Is the service called tools? Either way please make it italics.

File: README.md [hunk 13]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+- The Windows image has been sysprep'ed and it is available for cloning using the following command:
+`C:\Windows\System32\Sysprep\sysprep.exe /generalize /oobe /shutdown /unattend:<path_to_unattend.xml>`
+
+Sample [unattend.xml](https://gist.github.com/ravisantoshgudimetla/9d8e26032810a8df0b71b8d67f4d906b)
```
Comment (aravindhp) [inline README.md:28]: Please add this file to the repo

File: README.md [hunk 14]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+#### Windows VM assumptions:
+- An updated version of Windows Server 1909 VM image and corresponding container images are used
+- The Windows image is configured with SSH so that WMCO can access it:
+  - SSH on the Windows VM is configured using the [powershell script](https://gist.githubusercontent.com/ravisantoshgudimetla/087bf49088f70edf7f97777386d12f18/raw/0f13048697130a0265ed1050591361fdebcee3ae/powershell.yaml)
+  - A new file called C:\Users\Administrator\.ssh\authorized_keys containing the public key to be used is created
+- VMWare tools is installed, configured and running on the Windows VM:
+  - The [tools.conf](https://docs.vmware.com/en/VMware-Tools/11.1.0/com.vmware.vsphere.vmwaretools.doc/GUID-EA16729B-43C9-4DF9-B780-9B358E71B4AB.html) file at C:\ProgramData\VMware\VMware Tools\tools.conf has the following entry to ensure that the cloned vNIC generated on the Windows VM by hybrid-overlay won't be ignored and the VM has an IP Address in the vCenter
+    - `exclude-nics=`
+  - The tools Windows service is running in the Windows VM.
+- The Windows image has been sysprep'ed and it is available for cloning using the following command:
```
Comment (aravindhp) [inline README.md:25]: `The Windows image has been sysperp'ed and made ready for clone using the following command:`

File: README.md [hunk 15]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+`C:\Windows\System32\Sysprep\sysprep.exe /generalize /oobe /shutdown /unattend:<path_to_unattend.xml>`
+
+Sample [unattend.xml](https://gist.github.com/ravisantoshgudimetla/9d8e26032810a8df0b71b8d67f4d906b)
+
+
```
Comment (aravindhp) [inline README.md:30]: Please remove extra line

File: README.md [hunk 16]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+- Generally, ignition config files are available on the internal API Server URL in case of IPI however in case of
```
Comment (aravindhp) [inline README.md:31]: Please indent rest of the paragraph after the `-`

`Generally with IPI installs, ignition config files are available on the internal API server URL. However in the case of vSphere...`

File: README.md [hunk 17]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+    - `exclude-nics=`
+  - The tools Windows service is running in the Windows VM.
+- The Windows image has been sysprep'ed and it is available for cloning using the following command:
+`C:\Windows\System32\Sysprep\sysprep.exe /generalize /oobe /shutdown /unattend:<path_to_unattend.xml>`
+
+Sample [unattend.xml](https://gist.github.com/ravisantoshgudimetla/9d8e26032810a8df0b71b8d67f4d906b)
+
+
+- Generally, ignition config files are available on the internal API Server URL in case of IPI however in case of 
+vSphere, the ignition files are available on both internal and external API Server URLs. WMCO as of now, downloads the
```
Comment (aravindhp) [inline README.md:32]: `WMCO downloads...`. No need for `As of now` as it is automatically implied.

File: README.md [hunk 18]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+ignition files from internal API server. So, add a new DNS entry for `api-int.<cluster-name>.domain.com` which points
```
Comment (aravindhp) [inline README.md:33]: `the internal...`. Please explain why this DNS entry is needed.

File: README.md [hunk 19]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+to external API server URL `api.<cluster-name>.domain.com`. The external api should have been created as part of 
+[cluster install](https://docs.openshift.com/container-platform/4.5/installing/installing_vsphere/installing-vsphere-installer-provisioned.html)
+This requirement will be removed in the future.
```
Comment (aravindhp) [inline README.md:36]: Have you created a story for this?

File: README.md [hunk 20]
```diff
@@ -12,6 +12,35 @@ More design details can be explored in the [WMCO enhancement](https://github.com
...
+
+vSphere Windows VMs are cloned off an existing template or VM, for every VM the following steps need to be done. It is assumed that sysprep has been run on the VM:
+- Change the hostname of the Windows VM to match Windows Machine you created. The following powershell command can be used to do so:
+```
+Rename-Computer -NewName <machine_name> -Force -Restart
+```
```
Comment (aravindhp) [inline README.md:42]: Isn't WMCO doing this step?

File: README.md [hunk 21]
```diff
@@ -36,6 +65,8 @@ the MachineSet should have following labels:
 * *machine.openshift.io/cluster-api-machine-role: worker*
 * *machine.openshift.io/cluster-api-machine-type: worker*
 
+### Creating a vSphere Windows MachineSet
+
```
Comment (aravindhp) [inline README.md:69]: Please remove

File: README.md [hunk 22]
```diff
@@ -46,11 +77,7 @@ Not having these labels will result in the Windows node not being marked as a wo
 oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
 ```
 
-`<location>` should be replaced with a valid Azure location like `centralus`.
-
-`<zone>` should be replaced with a valid Azure availability zone like `us-east-1a`.
-
-Please note that on Azure, Windows Machine names cannot be more than 15 characters long.
+Please note that on vSphere Windows Machine names cannot be more than 15 characters long.
```
Comment (aravindhp) [inline README.md:80]: `vSphere, Windows Machine`.

So this restriction applies to vSphere also.

File: README.md [hunk 23]
```diff
@@ -80,37 +107,35 @@ spec:
...
+          memoryMiB: 16384
+          metadata:
+            creationTimestamp: null
+          network:
+            devices:
+            - networkName: dev-segment
+          numCPUs: 4
+          numCoresPerSocket: 1
+          snapshot: ""
+          template: 1909-template-docker-ssh
```
Comment (aravindhp) [inline README.md:124]: This requires an explanation

File: README.md [hunk 24]
```diff
@@ -80,37 +107,35 @@ spec:
...
-            namespace: openshift-machine-api
-          vmSize: Standard_D2s_v3
-          vnet: <infrastructureID>-vnet
-          zone: "<zone>"
+            name: worker-user-data-managed
+          workspace:
+            datacenter: xyz
+            datastore: abc
+            folder: folder_abc
+            resourcePool: resource_abc
```
Comment (aravindhp) [inline README.md:131]: Needs an explanation. Please use `dc_name` and `ds_name` and so on

File: docs/machineset-azure.md [hunk 25]
```diff
@@ -0,0 +1,70 @@
+# Creating an Azure Windows MachineSet
+
+_\<infrastructureID\>_ should be replaced with the output of:
+```shell script
+ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
+```
+
+<location> should be replaced with a valid Azure location like centralus.
```
Comment (aravindhp) [inline docs/machineset-azure.md:8]: Please follow the same formatting in [machineset-aws.md](https://github.com/openshift/windows-machine-config-operator/blob/master/docs/machineset-aws.md) through this doc

File: docs/machineset-azure.md [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+
+<zone> should be replaced with a valid Azure availability zone like us-east-1a.
+
+Please note that on Azure, Windows Machine names cannot be more than 15 characters long. The MachineSet name can therefore not be more than 9 characters long, due to the way Machine names are generated from it.
```
Comment (aravindhp) [inline docs/machineset-azure.md:12]: 120 column

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @ravisantoshgudimetla. Please address my comments and move this into #139

**Comment by ravisantoshgudimetla** [comment]
/close

folded into #139

**Pull request closed without merging**